repos/fdbt-site/src/pages/api/multipleProducts.ts

```diff
@@ -146,38 +169,139 @@ export const checkProductNamesAreValid = (products: MultiProduct[]): MultiProduc
+        validationResult = checkCarnetExpiriesAreValid(validationResult);
+        validationResult = checkCarnetExpiryUnitsAreValid(validationResult);
+    }
+
+    return validationResult;
+};
+
 export default (req: NextApiRequestWithSession, res: NextApiResponse): void => {
     try {
-        const numberOfProducts = Number(
-            (getSessionAttribute(req, NUMBER_OF_PRODUCTS_ATTRIBUTE) as NumberOfProductsAttribute).numberOfProductsInput,
-        );
+        const carnetAttribute = getSessionAttribute(req, CARNET_FARE_TYPE_ATTRIBUTE);
+        const isCarnet: boolean = carnetAttribute !== undefined && !!carnetAttribute;
+        const fareTypeAttribute = getSessionAttribute(req, FARE_TYPE_ATTRIBUTE);
+        if (!fareTypeAttribute || isFareTypeAttributeWithErrors(fareTypeAttribute)) {
+            throw new Error('Faretype attribute not found, could not ascertain fareType.');
+        }
+        const schoolFareType = getSessionAttribute(req, SCHOOL_FARE_TYPE_ATTRIBUTE);
+        const isFlatFare =
+            fareTypeAttribute.fareType === 'flatFare' ||
+            (fareTypeAttribute.fareType === 'schoolService' && schoolFareType?.schoolFareType === 'flatFare');
         const multipleProducts: MultiProduct[] = [];
-        for (let i = 0; i < numberOfProducts; i += 1) {
+        let i = 0;
+        while (req.body[`multipleProductNameInput${i}`] !== undefined) {
             const productName = req.body[`multipleProductNameInput${i}`];
             const productPrice = req.body[`multipleProductPriceInput${i}`];
             const productDuration = req.body[`multipleProductDurationInput${i}`];
-            const productDurationUnits = req.body[`multipleProductDurationUnitsInput${i}`] || '';
+            const productDurationUnits = req.body[`multipleProductDurationUnitsInput${i}`];
+            const quantity = req.body[`carnetQuantityInput${i}`];
+            const expiryTime = req.body[`carnetExpiryDurationInput${i}`];
+            const expiryUnit = req.body[`carnetExpiryUnitInput${i}`];
             const productNameId = `multiple-product-name-${i}`;
             const productPriceId = `multiple-product-price-${i}`;
-            const productDurationId = `multiple-product-duration-${i}`;
-            const productDurationUnitsId = `multiple-product-duration-units-${i}`;
+            const productDurationId = `product-details-period-duration-quantity-${i}`;
+            const productDurationUnitsId = `product-details-period-duration-unit-${i}`;
+            const productCarnetQuantityId = `product-details-carnet-quantity-${i}`;
+            const productCarnetExpiryDurationId = `product-details-carnet-expiry-quantity-${i}`;
+            const productCarnetExpiryUnitsId = `product-details-carnet-expiry-unit-${i}`;
             const product: MultiProduct = {
                 productName,
                 productNameId,
                 productPrice,
                 productPriceId,
+                carnetDetails: quantity && {
```

Comment:
This is a check to ensure the carnetDetails exist (we assume if there is quantity, the other ones will be there too. The reason we do this is so that the object is undefined if quantity isnt here.